src/transaction/confirm.rs

```diff
@@ -79,228 +70,190 @@ impl Default for ConfirmParams {
-                        let starting_block_num = *starting_block_num.get_or_insert(block_num);
-                        let elapsed_blocks = block_num.saturating_sub(starting_block_num);
-
-                        if elapsed_blocks > U64::from(block_timeout) {
-                            return Poll::Ready(Err(ExecutionError::ConfirmTimeout));
-                        }
+                        return Ok(Check::Confirmed(tx));
                     }
 
-                    ConfirmState::CreatingFilter(
-                        web3.eth_filter().create_blocks_filter().compat(),
-                        target_block_num,
-                        remaining_confirmations.as_u64(),
-                    )
-                }
-                ConfirmState::CreatingFilter(create_filter, target_block_num, count) => {
-                    match ready!(create_filter.poll(cx)) {
-                        Ok(filter) => ConfirmState::WaitingForBlocks(
-                            filter
-                                .stream(params.poll_interval)
-                                .skip(*count - 1)
-                                .into_future()
-                                .compat(),
-                            *target_block_num,
-                        ),
-                        Err(_) => {
-                            // NOTE: In the case we fail to create a filter
-                            //   (usually because the node doesn't support
-                            //   filters like Infura over HTTPS) then fall back
-                            //   to polling.
-                            ConfirmState::PollDelay(delay(params.poll_interval), *target_block_num)
-                        }
-                    }
-                }
-                ConfirmState::WaitingForBlocks(wait, target_block_num) => {
-                    match ready!(wait.poll(cx)) {
-                        Ok(_) => ConfirmState::Check,
-                        Err(_) => {
-                            // NOTE: In the case we fail to query the filter
-                            //   (node is behind a load balancer or cleaned up
-                            //   the filter) then also fall back to polling.
-                            ConfirmState::PollDelay(delay(params.poll_interval), *target_block_num)
-                        }
-                    }
+                    (target_block, remaining_confirmations.as_usize())
                 }
-                ConfirmState::PollDelay(delay, target_block_num) => {
-                    ready!(delay.poll(cx));
-                    ConfirmState::PollCheckingBlockNumber(
-                        web3.eth().block_number().compat(),
-                        *target_block_num,
+                None => {
+                    let remaining_confirmations = self.params.confirmations + 1;
+                    (
+                        latest_block + remaining_confirmations,
```

Comment:
Shouldn't this be lower bound by the confirmation timeout or are we already asserting somewhere that the timeout is larger than the remaining confirmations?

Reply:
Yes! Nice catch - I think this was a bug in the previous implementation as well.

Reply:
I submitted a fix in #327 , I'll merge this PR as-is so that it has the exact same behaviour as before (so this PR is purely a state-machine -> async/await conversion).